src/modules/routing/services/MediatorService.ts

```diff
@@ -0,0 +1,241 @@
+      if (index > -1) {
+        mediationRecord.recipientKeys.splice(index, 1)
+      }
+      this.mediationRepository.update(mediationRecord)
+      return KeylistUpdateResult.Success
+    }
+    return KeylistUpdateResult.ServerError
+  }
+
+  public async findRecipient(recipientKey: Verkey): Promise<string | null> {
+    const records = await this.getAll()
+    for (const record of records) {
+      for (const key of record.recipientKeys) {
+        if (recipientKey == key) {
+          return record.connectionId
+        }
+      }
+    }
+    return null
+  }
+  // TODO: remove this in favor of find by connection id
+  public async findRecipientByConnectionId(connectionId: string): Promise<MediationRecord | null> {
+    try {
+      const records = await this.mediationRepository.findByQuery({ connectionId })
+      return records[0]
+    } catch (error) {
+      return null
+    }
+  }
+  // TODO: resolve possible duplicate keylist messages
+  //public async createKeylistUpdateResponseMessage(keylist: KeylistUpdated[]): Promise<KeylistUpdateResponseMessage> {
+  //const keylistUpdateMessage = new KeylistUpdateResponseMessage({
+  //  updated: keylist,
+  //})
+  //return keylistUpdateMessage
+  //}
+
+  public async createGrantMediationMessage(mediation: MediationRecord): Promise<MediationGrantMessage> {
+    if (this.routingKeys.length === 0) {
+      /* eslint-disable @typescript-eslint/no-unused-vars */
+      const [_, verkey] = await this.wallet.createDid()
+      /* eslint-enable @typescript-eslint/no-unused-vars */
+      this.routingKeys = [verkey]
+    }
+    mediation.state = MediationState.Granted
+    await this.mediationRepository.update(mediation)
+    return new MediationGrantMessage({
+      id: uuid(), // HELP: should this be the thread id from the request?
+      endpoint: this.agentConfig.getEndpoint(),
+      routingKeys: this.routingKeys,
+    })
+  }
+
+  public async processMediationRequest(messageContext: InboundMessageContext<MediationRequestMessage>) {
+    // Assert connection
+    const connection = this._assertConnection(messageContext.connection, MediationRequestMessage)
+
+    const mediationRecord = await createRecord(
+      {
+        connectionId: connection.id,
+        role: MediationRole.Mediator,
+        state: MediationState.Init,
+      },
+      this.mediationRepository
+    )
+    await this.updateState(mediationRecord, MediationState.Init)
+
+    const message = await this.createGrantMediationMessage(mediationRecord)
+    const outbound = createOutboundMessage(connection, message)
+    if (messageContext.message.hasReturnRouting()) {
+      return outbound
+    } else {
+      await this.messageSender.sendMessage(outbound)
+    }
+  }
+
+  public async findByConnectionId(connectionId: string): Promise<MediationRecord | null> {
+    try {
+      const records = await this.mediationRepository.findByQuery({ connectionId })
+      return records[0]
+    } catch (error) {
+      return null
```

Comment:
Not sure if it's a good idea to silent the error. We should add at least some log.

Reply:
adding 
```ts
    } catch (error) {
      this.agentConfig.logger.warn(`Could not find record for connection: ${connectionId}`)
      return null
    }
```